File: src/main/java/com/thonbecker/BibleVerseStack.java
package com.thonbecker;

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;

import software.amazon.awscdk.core.BundlingOptions;
import software.amazon.awscdk.core.Construct;
import software.amazon.awscdk.core.DockerVolume;
import software.amazon.awscdk.core.Duration;
import software.amazon.awscdk.core.Stack;
import software.amazon.awscdk.core.StackProps;
import software.amazon.awscdk.core.RemovalPolicy;
import software.amazon.awscdk.services.s3.Bucket;
import software.amazon.awscdk.services.s3.BucketProps;
import software.amazon.awscdk.services.s3.deployment.BucketDeployment;
import software.amazon.awscdk.services.s3.deployment.BucketDeploymentProps;
import software.amazon.awscdk.services.s3.deployment.ISource;
import software.amazon.awscdk.services.s3.deployment.Source;
import software.amazon.awscdk.services.s3.assets.AssetOptions;
import software.amazon.awscdk.services.apigatewayv2.CorsPreflightOptions;
import software.amazon.awscdk.services.apigatewayv2.DomainMappingOptions;
import software.amazon.awscdk.services.apigatewayv2.DomainName;
import software.amazon.awscdk.services.apigatewayv2.DomainNameProps;
import software.amazon.awscdk.services.apigatewayv2.HttpApi;
import software.amazon.awscdk.services.apigatewayv2.AddRoutesOptions;
import software.amazon.awscdk.services.apigatewayv2.CorsHttpMethod;
import software.amazon.awscdk.services.apigatewayv2.HttpApiProps;
import software.amazon.awscdk.services.apigatewayv2.HttpMethod;
import software.amazon.awscdk.services.apigatewayv2.PayloadFormatVersion;
import software.amazon.awscdk.services.apigatewayv2.integrations.HttpLambdaIntegration;
import software.amazon.awscdk.services.apigatewayv2.integrations.HttpLambdaIntegrationProps;
import software.amazon.awscdk.services.certificatemanager.Certificate;
import software.amazon.awscdk.services.route53.HostedZone;
import software.amazon.awscdk.services.route53.HostedZoneProviderProps;
import software.amazon.awscdk.services.route53.ARecord;
import software.amazon.awscdk.services.route53.ARecordProps;
import software.amazon.awscdk.services.route53.RecordTarget;
import software.amazon.awscdk.services.route53.targets.ApiGatewayv2DomainProperties;
import software.amazon.awscdk.services.lambda.Code;
import software.amazon.awscdk.services.lambda.Function;
import software.amazon.awscdk.services.lambda.FunctionProps;
import software.amazon.awscdk.services.lambda.Runtime;
import software.amazon.awscdk.services.logs.RetentionDays;

import static java.util.Collections.singletonList;
import static software.amazon.awscdk.core.BundlingOutput.ARCHIVED;

public class BibleVerseStack extends Stack {

    public BibleVerseStack(final Construct scope, final String id, final StackProps props) {
        super(scope, id, props);

        Bucket bucket = new Bucket(this, "Bucket", new BucketProps.Builder()
                .removalPolicy(RemovalPolicy.DESTROY).autoDeleteObjects(true).build());

        List<ISource> sources = new ArrayList<>(1);
        sources.add(Source.asset("./data"));

        BucketDeployment bucketDeployment = new BucketDeployment(this, "BucketDeployment",
                new BucketDeploymentProps.Builder().sources(sources).destinationBucket(bucket).build());

        List<String> BibleVersePackagingInstructions = Arrays.asList("/bin/sh", "-c",
                "mvn clean install " + "&& cp /asset-input/target/bibleverse.jar /asset-output/");

        BundlingOptions.Builder builderOptions = BundlingOptions.builder()
                .command(BibleVersePackagingInstructions).image(Runtime.JAVA_11.getBundlingImage())
                .volumes(singletonList(DockerVolume.builder()
                        .hostPath(System.getProperty("user.home") + "/.m2/")
                        .containerPath("/root/.m2/").build()))
                .user("root").outputType(ARCHIVED);

        Map<String, String> environmentMap = new HashMap<>();
        environmentMap.put("DATA_BUCKET_NAME", bucket.getBucketName());

        Function functionRandomBibleVerse = new Function(this, "FunctionRandomBibleVerse", FunctionProps
                .builder().runtime(Runtime.JAVA_11)
                .code(Code.fromAsset("./lambda/",
                        AssetOptions.builder().bundling(builderOptions.build()).build()))
                .handler("com.thonbecker.randombibleverse.LambdaHandler").memorySize(1024)
                .timeout(Duration.seconds(10)).logRetention(RetentionDays.ONE_WEEK)
                .environment(environmentMap).build());

        Function functionAboutBibleVerse = new Function(this, "FunctionAboutBibleVerse", FunctionProps.builder()
                .runtime(Runtime.JAVA_11)
                .code(Code.fromAsset("./lambda/",
                        AssetOptions.builder().bundling(builderOptions.build()).build()))
                .handler("com.thonbecker.about.LambdaHandler").memorySize(1024)
                .timeout(Duration.seconds(10)).logRetention(RetentionDays.ONE_WEEK)
                .environment(environmentMap).build());

        bucket.grantRead(functionRandomBibleVerse);

        DomainName domainName = new DomainName(this, "DomainName", DomainNameProps.builder()
                .domainName("bibleverse.thonbecker.com")
                .certificate(Certificate.fromCertificateArn(this, "CertificateLookup",
                        "arn:aws:acm:us-east-1:664759038511:certificate/1554896f-abd0-4a6b-8955-710f2f4a642b"))
                .build());

        List<String> corsOrigins = new ArrayList<>(1);
        corsOrigins.add("*");

        List<CorsHttpMethod> corsHttpMethods = new ArrayList<>(2);
        corsHttpMethods.add(CorsHttpMethod.GET);
        corsHttpMethods.add(CorsHttpMethod.OPTIONS);

        List<HttpMethod> httpMethods = new ArrayList<>();
        httpMethods.add(HttpMethod.GET);

        HttpApi httpApi = new HttpApi(this, "bibleverse-api", HttpApiProps.builder().apiName("bibleverse-api")
                .createDefaultStage(true)
                .corsPreflight(CorsPreflightOptions.builder().allowMethods(corsHttpMethods)
                        .allowOrigins(corsOrigins).build())
                .defaultDomainMapping(DomainMappingOptions.builder().domainName(domainName).build())
                .build());

        httpApi.addRoutes(AddRoutesOptions.builder().path("/about").methods(httpMethods)
                .integration(new HttpLambdaIntegration("LambdaIntegrationAbout", functionAboutBibleVerse,
                HttpLambdaIntegrationProps.builder()
                        .payloadFormatVersion(PayloadFormatVersion.VERSION_2_0).build()))
                .build());

        httpApi.addRoutes(AddRoutesOptions.builder().path("/random").methods(httpMethods)
                .integration(new HttpLambdaIntegration("LambdaIntegrationRandom",functionRandomBibleVerse,
                HttpLambdaIntegrationProps.builder()
                        .payloadFormatVersion(PayloadFormatVersion.VERSION_2_0).build()))
                .build());

        ARecord aliasRecord = new ARecord(this, "AliasRecord", ARecordProps.builder()
                .recordName("bibleverse.thonbecker.com")
                .zone(HostedZone.fromLookup(this, "HostedZoneLookup",
                        HostedZoneProviderProps.builder().domainName("thonbecker.com").build()))
                .target(RecordTarget.fromAlias(
                        new ApiGatewayv2DomainProperties(domainName.getRegionalDomainName(),
                                domainName.getRegionalHostedZoneId())))
                .build());
    }
}

File: .infrastructure/lib/bibleverse-stack.ts
import { Stack, StackProps, RemovalPolicy, Duration, } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as s3deploy from 'aws-cdk-lib/aws-s3-deployment';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as acm from 'aws-cdk-lib/aws-certificatemanager';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as targets from 'aws-cdk-lib/aws-route53-targets';
import {
  HttpApi,
  DomainName,
  HttpMethod,
  CorsHttpMethod,
} from 'aws-cdk-lib/aws-apigatewayv2';
import {
  HttpLambdaIntegration,
} from 'aws-cdk-lib/aws-apigatewayv2-integrations';
import * as path from "path";
import * as logs from 'aws-cdk-lib/aws-logs';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';

export class BibleVerseStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const certArn = 'arn:aws:acm:us-east-1:664759038511:certificate/1554896f-abd0-4a6b-8955-710f2f4a642b';
    const domainName = 'bibleverse.thonbecker.com';

    // Create log groups for Lambda functions
    const aboutFunctionLogGroup = new logs.LogGroup(this, 'AboutFunctionLogGroup', {
      logGroupName: '/aws/lambda/BibleVerseStack-AboutFunction',
      retention: RetentionDays.THREE_DAYS,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const randomBibleVerseFunctionLogGroup = new logs.LogGroup(this, 'RandomBibleVerseFunctionLogGroup', {
      logGroupName: '/aws/lambda/BibleVerseStack-RandomBibleVerseFunction',
      retention: RetentionDays.THREE_DAYS,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const aboutFunction = new lambda.Function(this, 'AboutFunction', {
      runtime: lambda.Runtime.JAVA_21,
      memorySize: 1024,
      timeout: Duration.seconds(30),
      logGroup: aboutFunctionLogGroup,
      handler: 'org.springframework.cloud.function.adapter.aws.FunctionInvoker',
      code: lambda.Code.fromAsset('../target/bibleverse-2.0.0-aws.jar'),
      environment: {
        'SPRING_CLOUD_FUNCTION_DEFINITION': 'aboutHandler'
      },
    });

    const randomBibleVerseFunction = new lambda.Function(this, 'RandomBibleVerseFunction', {
      runtime: lambda.Runtime.JAVA_21,
      memorySize: 1024,
      timeout: Duration.seconds(30),
      logGroup: randomBibleVerseFunctionLogGroup,
      handler: 'org.springframework.cloud.function.adapter.aws.FunctionInvoker',
      code: lambda.Code.fromAsset('../target/bibleverse-2.0.0-aws.jar'),
      environment: {
        'SPRING_CLOUD_FUNCTION_DEFINITION': 'randomBibleVerseHandler'
      },
    });

    const bucket = new s3.Bucket(this, 'BibleVerseBucket', {
      bucketName: 'bible-verse-data-files',
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_ENFORCED,
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const deployment = new s3deploy.BucketDeployment(this, 'DeployBibleVerse', {
      sources: [s3deploy.Source.asset(path.join(__dirname, 'data', 'files'))],
      destinationBucket: bucket,
    });

    deployment.deployedBucket.grantRead(randomBibleVerseFunction);
    
    const dn = new DomainName(this, 'DomainNameBibleVerse', {
      domainName,
      certificate: acm.Certificate.fromCertificateArn(this, 'CertificateBibleVerse', certArn),
    });

    const api = new HttpApi(this, 'HttpApiBibleVerse', {
      corsPreflight: {
        allowOrigins: ['*'],
        allowMethods: [CorsHttpMethod.GET, CorsHttpMethod.OPTIONS],
        allowHeaders: [
          'Content-Type',
          'Access-Control-Allow-Headers',
          'Access-Control-Allow-Origin',
          'Access-Control-Allow-Methods',
        ],
      },
      defaultIntegration: new HttpLambdaIntegration('DefaultIntegration', randomBibleVerseFunction),
      defaultDomainMapping: {
        domainName: dn,
      },
    });

    api.addRoutes({
      path: '/about',
      methods: [HttpMethod.GET],
      integration: new HttpLambdaIntegration('AboutIntegration', aboutFunction)
    });

    new route53.ARecord(this, 'AliasBibleVerse', {
      zone: route53.HostedZone.fromHostedZoneAttributes(this, 'HostedZoneBibleVerse', {
        hostedZoneId: 'Z0960080GF0UBO75OWWP',
        zoneName: 'thonbecker.com'
      }),
      recordName: 'bibleverse.thonbecker.com',
      target: route53.RecordTarget.fromAlias(new targets.ApiGatewayv2DomainProperties(dn.regionalDomainName, dn.regionalHostedZoneId)),
    });
  }
}
